File: code/primes.scala
// compute first n primes
def primes(n: Int): List[Int] = {
  def isPrime(n: Int, primes: List[Int]): Boolean = {
    primes.forall(n % _ != 0)
  }
  def loop(n: Int, primes: List[Int]): List[Int] = {
    if (primes.length == n) primes
    else if (isPrime(primes.length + 1, primes)) loop(n, primes :+ (primes.length + 1))
    else loop(n, primes)
  }
  loop(n, List(2))
}

// eratosthenes siege to compute first n primes using functional programming
def eratosthenes(n: Int): List[Int] = {
  def sieve(n: Int, primes: List[Int]): List[Int] = {
    if (primes.length == n) primes
    else sieve(n, primes.filter(_ % primes.length != 0))
  }
  sieve(n, List.range(1, n))
}

File: code/primes.js
/* check if a number is prime */
function isPrime(n) {
  if (n < 2) return false;
  for (var i = 2; i < n; i++) {
    if (n % i == 0) return false;
  }
  return true;
}

/* Erathostenes siege */
function eratosthenes(n) {
  var primes = [];
  for (var i = 2; i <= n; i++) {
    if (isPrime(i)) primes.push(i);
  }
  return primes;
}

/* Eratosthenes sieve with a bit array */
function eratosthenesBit(n) {
  var primes = [];
  var bits = new Array(n + 1);
  for (var i = 2; i <= n; i++) {
    if (!bits[i]) {
      primes.push(i);
      for (var j = i; j <= n; j += i) {
        bits[j] = true;
      }
    }
  }
  return primes;
}